src/parser/parse_string.ts

```diff
@@ -117,3 +124,423 @@ function decodeEscape(p: Parser, s: string) {
+                   chars (not part of 2-char tokens). If by themselves, they
+                   don't end an expression (unlike say '!'). */
+                if (ch === ">" || ch === "<") {
+                    continue;
+                }
+            }
+
+            /* Normal way out of this loop. */
+            break;
+        } else {
+            /* Just consume this char and loop around. */
+        }
+    }
+
+    /* If we leave this loop in a string or with mismatched parens, we
+       don't care. We'll get a syntax error when compiling the
+       expression. But, we can produce a better error message, so
+       let's just do that.*/
+    if (quote_char) {
+        p.raise_error(pySyntaxError, "f-string: unterminated string");
+    }
+    if (nested_depth) {
+        p.raise_error(pySyntaxError, "f-string: mismatched '(', '{', or '['");
+    }
+
+    const expr_end = i;
+
+    /* Compile the expression as soon as possible, so we show errors
+       related to the expression before errors related to the
+       conversion or format_spec. */
+    const simple_expression = fstring_compile_expr(p, str, expr_start, expr_end, t);
+
+    /* Check for =, which puts the text value of the expression in
+       expr_text. */
+    if (str[i] === "=") {
+        i++;
+        /* Skip over ASCII whitespace.  No need to test for end of string
+           here, since we know there's at least a trailing quote somewhere
+           ahead. */
+        while (WHITE_SPACE.test(str[i])) {
+            i++;
+        }
+        expr_text = str.slice(expr_start, i);
+    }
+
+    /* Check for a conversion char, if present. */
+    if (str[i] === "!") {
+        i++;
+        if (i >= end) unexpected_end_of_string(p);
+
+        conversion = str[i];
+        i++;
+
+        /* Validate the conversion. */
+        if (!(conversion === "s" || conversion === "r" || conversion === "a")) {
+            p.raise_error(pySyntaxError, "f-string: invalid conversion character: expected 's', 'r', or 'a'");
+        }
+    }
+
+    /* Check for the format spec, if present. */
+    if (i >= end) unexpected_end_of_string(p);
+    if (str[i] === ":") {
+        i++;
+        if (i >= end) unexpected_end_of_string(p);
+        /* Parse the format spec. */
+        [format_spec, i] = fstring_parse(p, str, i, end, raw, recurse_lvl + 1, this.first, t, this.last);
+    }
+
+    if (i >= end || str[i] !== "}") unexpected_end_of_string(p);
+
+    /* We're at a right brace. Consume it. */
+    i++;
+
+    if (expr_text && format_spec === null && conversion === null) {
+        conversion = "r";
+    }
+
+    /* And now create the FormattedValue node that represents this
+       entire expression with the conversion and format spec. */
+    const expr = new FormattedValue(
+        simple_expression,
+        /** @todo change this to accept "r", "s", "a" but to match cpython ast we'll need to output the charCode */
+        conversion === null ? -1 : conversion.charCodeAt(0),
```

Comment:
python reports the charCode for conversions
skulpt uses the char

we either hack the AST dump and adl patch
or 
change skulpt to convert the `conversion` field to a string

Thoughts?

Reply:
I don't know lets have a chat about this 😄 it might also be 11pm too much rn